js/src/logger.ts

```diff
@@ -406,15 +413,144 @@ interface UserInfo {
   id: string;
 }
 
-export class Project {
-  name: string;
+interface RegisteredProject {
   id: string;
-  org_id: string;
+  name: string;
+}
 
-  constructor(name: string, id: string, org_id: string) {
+export class Project {
+  name?: string;
+  id?: string;
+
+  constructor({ name, id }: { name?: string; id?: string }) {
     this.name = name;
     this.id = id;
-    this.org_id = org_id;
+  }
+
+  async lazyInit(): Promise<RegisteredProject> {
+    if (this.id === undefined) {
+      const response = await _state
+        .apiConn()
+        .post_json("api/project/register", {
+          project_name: this.name || GLOBAL_PROJECT,
+          org_id: _state.orgId,
+        });
+      this.id = response.project.id;
+      if (this.name === undefined) {
+        this.name = response.project.name;
+      }
+    } else if (this.name === undefined) {
+      const response = await _state.apiConn().get_json("api/project", {
+        id: this.id,
+      });
+      this.name = response.name;
+    }
+
+    return { id: this.id!, name: this.name! };
+  }
+}
+
+export interface LogOptions {
+  is_async?: boolean;
+}
+
+export class Logger {
+  lazyLogin: () => Promise<void>;
+  lazyProject: Project;
+  logOptions: LogOptions;
+  private logger: LogThread;
+  private lastStartTime: number;
+
+  constructor(
+    lazyLogin: () => Promise<void>,
+    lazyProject: Project,
+    logOptions: LogOptions = {}
+  ) {
+    this.lazyLogin = lazyLogin;
+    this.lazyProject = lazyProject;
+    this.logOptions = logOptions;
+
+    this.logger = new LogThread();
+    this.lastStartTime = getCurrentUnixTimestamp();
```

Comment:
fixed by adding a `lazyLogin()` method that resets it

Reply:
I think the solution would have to be more tricky than this, because now we're not accounting for the work done leading up to the first log statement (i.e. the first span duration will effectively be 0).

One way to do it "correctly" is to time how long `lazyLogin` took and add that to the `startSpan` start time. But that only fixes the relative duration at the expense of a meaningful absolute start time.

Reply:
discussed offline: we expect the common case to be `traced()` and loggers to be initialized far earlier than the first request coming in, so this tradeoff works